homeassistant/components/pi_hole/__init__.py

```diff
@@ -16,95 +16,172 @@
+        location = conf[CONF_LOCATION]
+        api_key = conf.get(CONF_API_KEY)
+
+        LOGGER.debug("Setting up %s integration with host %s", DOMAIN, host)
+
+        session = async_get_clientsession(hass, verify_tls)
+        pi_hole = PiHoleData(
+            Hole(
+                host,
+                hass.loop,
+                session,
+                location=location,
+                tls=use_tls,
+                api_token=api_key,
+            ),
+            name,
+        )
 
-    conf = config[DOMAIN]
-    name = conf[CONF_NAME]
-    host = conf[CONF_HOST]
-    use_tls = conf[CONF_SSL]
-    verify_tls = conf[CONF_VERIFY_SSL]
-    location = conf[CONF_LOCATION]
-    api_key = conf.get(CONF_API_KEY)
-
-    LOGGER.debug("Setting up %s integration with host %s", DOMAIN, host)
-
-    session = async_get_clientsession(hass, verify_tls)
-    pi_hole = PiHoleData(
-        Hole(
-            host, hass.loop, session, location=location, tls=use_tls, api_token=api_key
-        ),
-        name,
-    )
-
-    await pi_hole.async_update()
-
-    hass.data[DOMAIN] = pi_hole
+        await pi_hole.async_update()
 
-    async def handle_disable(call):
-        if api_key is None:
-            raise vol.Invalid("Pi-hole api_key must be provided in configuration")
+        hass.data[DOMAIN][slugify(name)] = pi_hole
 
+    async def disable_service_handler(call):
+        """Handle the service call to disable a single Pi-Hole or all configured Pi-Holes."""
         duration = call.data[SERVICE_DISABLE_ATTR_DURATION].total_seconds()
-
-        LOGGER.debug("Disabling %s %s for %d seconds", DOMAIN, host, duration)
-        await pi_hole.api.disable(duration)
-
-    async def handle_enable(call):
-        if api_key is None:
-            raise vol.Invalid("Pi-hole api_key must be provided in configuration")
-
-        LOGGER.debug("Enabling %s %s", DOMAIN, host)
-        await pi_hole.api.enable()
+        name = call.data.get(SERVICE_DISABLE_ATTR_NAME)
+
+        async def do_disable(name):
+            """Disable the named Pi-Hole."""
+            slug = slugify(name)
+            if slug not in hass.data[DOMAIN]:
+                raise vol.Invalid(
+                    "Pi-hole '{}' not found. Check your configuration.".format(name)
+                )
+
+            pi_hole = hass.data[DOMAIN][slug]
+            if pi_hole.api.api_token is None:
+                raise vol.Invalid(
+                    "Pi-hole '{}' must have an api_key provided in configuration to be disabled.".format(
+                        pi_hole.name
+                    )
+                )
+
+            LOGGER.debug(
+                "Disabling Pi-hole '%s' (%s) for %d seconds",
+                name,
+                pi_hole.api.host,
+                duration,
+            )
+            await pi_hole.api.disable(duration)
+
+        if name is not None:
+            await do_disable(name)
+        else:
+            for pi_hole in hass.data[DOMAIN].values():
```

Comment:
I don't think we should allow disabling all in one service call.

Reply:
I disagree. My intent is to allow HA to provide an easy way for Pi-hole users to disable their Pi-hole, thus allowing unfiltered network access. The vast majority of users have only one Pi-hole, so it seems that requiring an additional parameter to the service call provides no value.

Reply:
Interesting because everyone I know using it runs two because a failure to DNS means your whole network goes down.  I already have a disable function setup to each pi hole via script today and then an input Boolean to disable both if needed. Sometimes I want to disable only 1 so I can force traffic for the other.  Having the disable traffic built into the service would be nice but I would want control to each device.

Reply:
Yeah, if we add the default name in the service schema the user doesn't need to set an additional parameter to disable a single Pi-hole. But if the user has more than one, it will be required to specify which one to disable. I think that makes sense.

Reply:
If the user has changed the name for the single Pi-hole the default will not work though. Unless we set the default to the single Pi-hole name? :shrug:

You can decide what works best as long as it's clear to the user and documented.